# Исследование надежности заемщиков


## Откройте таблицу и изучите общую информацию о данных

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

*Создаем сводную таблицу `data_children` группируя данные по количеству детей   
и вычисляя процент заемщиков, возвративщих кредит в срок*

In [22]:
# группируем датасет методом groupby по столбцу 'children', 
# в столбце 'count' выводим количество заемщиков по группам, 
# в столбце 'sum' - количество заемщиков, имевших задолженность по кредиту
data_children = data.groupby('children')['debt'].agg(['count', 'sum'])

# Добавляем столбец 'debt_percent' для подсчета процента заемщиков, имевших задолженность по кредиту
data_children['debt_percent'] = round(data_children['sum'] / data_children['count'] * 100, 2)  

In [23]:
borrowers = data['children'].count()  # Общее количество заемщиков

In [24]:
# Добавляем столбец 'debt_percent' для подсчета процента заемщиков, имевших задолженность по кредиту
data_children['borrowers_percent'] = round(data_children['count'] / borrowers * 100, 1)

In [25]:
data_children  

count   sum  debt_percent  borrowers_percent
children                                              
0         14091  1063          7.54               66.1
1          4808   444          9.23               22.5
2          2052   194          9.45                9.6
3           330    27          8.18                1.5
4            41     4          9.76                0.2
5             9     0          0.00                0.0

**Вывод:** **Бездетные заемщики реже имели задолженность по кредиту, чем заемщики, имеющие детей.** 

Семьи с одним или двумя детьми возвращали кредит более, чем в 90% случаев. 
Доля семей с тремя детьми, имевших задолженность по кредиту близка к доле бездетных семей, однако количество таких семей составляет всего 1,5% от всех заемщиков, поэтому эти данные не могут быть достоверными.  
Данных по заемщикам с четырьмя и пятью детьми недостаточно для корректных выводов по ним.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

*Создаем сводную таблицу `data_family_status`, группируя данные по семейному положению заемщика,  
вычисляем процент заемщиков, возвративщих кредит в срок*

In [26]:
# группируем датасет методом groupby по столбцу 'family_status', 
data_family_status = data.groupby('family_status')['debt'].agg(['count', 'sum'])

# Добавляем столбец 'percent' для подсчета процента заемщиков, имевших задолженность по кредиту
data_family_status['percent'] = round(data_family_status['sum'] / data_family_status['count'] * 100, 2)

In [27]:
data_family_status

count  sum  percent
family_status                             
Не женат / не замужем   2796  273     9.76
в разводе               1189   84     7.06
вдовец / вдова           951   63     6.62
гражданский брак        4134  385     9.31
женат / замужем        12261  927     7.56

**Вывод: Заемщики, которые состоят или когда-либо состояли в зарегистрированном браке надежнее тех, которые не женаты\не замужем.** 

Самая многочисленная группа заемщиков состоит в браке и имеет процент задолженности по кредиту 7,56%, незначительно превышающий процент задолженности заемщиков в разводе или вдовцов, т.е. тех, кто в прошлом состоял в браке - 7,06%. Наиболее надежными можно считать категорию заемщиков, являющихся вдовцами\вдовами. Процент задолженности по кредиту минимальный и составляет 6,62%. Вероятно людей, вступающих в зарегистрированный брак (в том числе в прошлом), можно считать наиболее ответственными, что положительно сказывается на их способности своевременно исполнять в том числе и финансовые обязательства.

Заемщики, состоящие в гражданском браке и не состоящие в браке имеют близкие значения процента задолженности: 9,31% и 9,76% соответственно. Самой ненадежной группой клиентов являются не женатые\не замужние.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [28]:
# группируем датасет методом groupby по столбцу 'total_income_category':
# 0–30000 — 'E';
# 30001–50000 — 'D';
# 50001–200000 — 'C';
# 200001–1000000 — 'B';
# 1000001 и выше — 'A'.

data_total_income_category = data.groupby('total_income_category')['debt'].agg(['count', 'sum'])

# Добавляем столбец 'percent' для подсчета процента заемщиков, имевших задолженность по кредиту
data_total_income_category['percent'] = round(data_total_income_category['sum'] / data_total_income_category['count'] * 100, 2)

In [29]:
data_total_income_category

count   sum  percent
total_income_category                      
A                         25     2     8.00
B                       5014   354     7.06
C                      15921  1353     8.50
D                        349    21     6.02
E                         22     2     9.09

*Количество заемщиков в категориях A и E настолько незначительно, что ими можно пренебречь. В то же время категория C слишком многочисленная, а разброс уровня дохода слишком большой. Требуется более детальная группировка данных по категориям.* 

*На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category1` с категориями:*

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–70000 — `'C6'`;
- 70001-90000 - `'C5'`;
- 90001-125000 - `'C4'`;
- 125001-150000 - `'C3'`;
- 150001-175000 - `'C2'`;
- 175001-200000 - `'C1'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


In [30]:
# создадим функцию, определяющую категорию заемщика по уровню дохода
def categorize_income1(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 70000:
            return 'C6'
        elif 70001 <= income <= 90000:
            return 'C5'
        elif 90001 <= income <= 125000:
            return 'C4'
        elif 125001 <= income <= 150000:
            return 'C3'
        elif 150001 <= income <= 175000:
            return 'C2'
        elif 175001 <= income <= 200000:
            return 'C1'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [31]:
data['total_income_category1'] = data['total_income'].apply(categorize_income1)

In [32]:
# группируем датасет методом groupby по столбцу 'total_income_category1':
data_total_income_category1 = data.groupby('total_income_category1')['debt'].agg(['count', 'sum'])

# Добавляем столбец 'percent' для подсчета процента заемщиков, имевших задолженность по кредиту
data_total_income_category1['percent'] = round(data_total_income_category1['sum'] / data_total_income_category1['count'] * 100, 2) 

In [33]:
data_total_income_category1

count  sum  percent
total_income_category1                     
A                          25    2     8.00
B                        5014  354     7.06
C1                       1704  149     8.74
C2                       3034  254     8.37
C3                       3838  331     8.62
C4                       4383  380     8.67
C5                       1869  161     8.61
C6                       1093   78     7.14
D                         349   21     6.02
E                          22    2     9.09

*Более детальная группировка заемщиков по категориям в диапазоне дохода 50001 - 200000 показала, что процент задолженности по кредиту существенно отличается только в категории С6 с доходом 50001 - 70000 рублей.* 

**Вывод: Чаще кредит в срок возвращают заемщики с доходом до 70000 рублей и доходом от 200001 до 1000000 рублей.  
Заемщики с доходом 70001 - 200000 рублей допускают задолженность по кредиту в среднем на 1,5% чаще предыдущей группы.  
Этот факт опровергает гипотезу, что клиент с большим доходом более надежный. Вероятно на возврат кредита в значительной степени влияют прочие факторы, и нельзя делать вывод о надежности клиента только по уровню дохода.**  
Количество заемщиков с доходом более 1000000 настолько незначительно, что невозможно сделать корректные выводы относительно их надежности.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [34]:
# группируем датасет методом groupby по столбцу 'purpose_category', 
data_purpose_category = data.groupby('purpose_category')['debt'].agg(['count', 'sum'])

# Добавляем столбец 'percent' для подсчета процента заемщиков, имевших задолженность по кредиту
data_purpose_category['percent'] = round(data_purpose_category['sum'] / data_purpose_category['count'] * 100, 2)

In [35]:
# Добавляем столбец 'purpose_percent' для подсчета процента заемщиков, имевших задолженность по кредиту
data_purpose_category['purpose_percent'] = round(data_purpose_category['count'] / borrowers * 100, 1)

In [36]:
data_purpose_category

count  sum  percent  purpose_percent
purpose_category                                              
операции с автомобилем     4279  400     9.35             20.1
операции с недвижимостью  10751  780     7.26             50.4
получение образования      3988  369     9.25             18.7
проведение свадьбы         2313  183     7.91             10.8

**Выясним цели кредита, которые наибольшим образом влияют на его возврат в срок:**

In [37]:
# группируем датасет методом groupby по столбцу 'purpose', 
data_purpose = data.groupby('purpose')['debt'].agg(['count', 'sum'])

# Добавляем столбец 'percent' для подсчета процента заемщиков, имевших задолженность по кредиту
data_purpose['percent'] = round(data_purpose['sum'] / data_purpose['count'] * 100, 2)

# Отсортируем датасет `data_purpose` по столбцу 'percent' в порядке убывания
data_purpose.sort_values(by='percent', ascending=False)

count  sum  percent
purpose                                                    
получение дополнительного образования     444   51    11.49
сделка с автомобилем                      455   50    10.99
получение высшего образования             425   46    10.82
сделка с подержанным автомобилем          481   51    10.60
профильное образование                    432   44    10.19
свой автомобиль                           473   47     9.94
заняться образованием                     408   39     9.56
на покупку автомобиля                     469   44     9.38
автомобили                                476   44     9.24
приобретение автомобиля                   459   42     9.15
на покупку своего автомобиля              504   46     9.13
высшее образование                        446   40     8.97
строительство недвижимости                619   54     8.72
заняться высшим образованием              496   43     8.67
дополнительное образование                455   38     8.35
автомобиль                                491   41     8.35
операции с недвижимостью                  672   55     8.18
получение образования                     440   36     8.18
свадьба                                   790   64     8.10
операции с коммерческой недвижимостью     645   52     8.06
операции со своей недвижимостью           623   50     8.03
покупка жилья для сдачи                   649   52     8.01
на проведение свадьбы                     763   61     7.99
строительство жилой недвижимости          620   48     7.74
сыграть свадьбу                           760   58     7.63
покупка жилья                             640   48     7.50
на покупку подержанного автомобиля        471   35     7.43
операции с жильем                         647   48     7.42
образование                               442   32     7.24
жилье                                     641   46     7.18
покупка жилья для семьи                   637   45     7.06
покупка коммерческой недвижимости         658   46     6.99
покупка недвижимости                      616   43     6.98
покупка жилой недвижимости                602   41     6.81
недвижимость                              631   42     6.66
строительство собственной недвижимости    628   41     6.53
ремонт жилью                              604   35     5.79
покупка своего жилья                      619   34     5.49

**Вывод: Возврат кредита в срок зависит от целей кредита.  
Кредиты на операции с недвижимостью более надежные, чем кредиты на операции с автомобилем и на получение образования.**  
  
Более половины заемщиков оформляют кредит на операции с недвижимостью. Процент таких заемщиков, имеющих задолженность по кредиту, минимальный и составляет 7,26%, что позволяет сделать вывод о наибольшей надежности этой категории заемщиков. К этой категории относится кредит на покупку своего жилья, имеющий самый низкий процент задолженности по кредиту (5,49%)  

Наиболее проблемной для целей возврата кредита является вторая по популярности группа заемщиков (20% всех займов), оформляющих кредит на операции с автомобилем. Процент таких заемщиков, имевших задолженность по кредиту, максимальный и составляет 9.35%. Вероятно в этом случае оправданным будет оформление страхования займов.  

Третья категория займов на получение образования (18,7% всех займов) имеет процент задолженности по кредиту 9.25% близкий к данным по категории 'операции с автомобилем'. Однако именно в этой категории присутствует наиболее проблемная группа заемщиков, оформивших кредит на получение дополнительного образования.	    

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:*  Пропущенные значения есть в двух столбцах:  
days_employed — общий трудовой стаж в днях;  
total_income — ежемесячный доход.  
Количество пропусков совпадает в этих столбцах и равно 2174 из 21331 строк датафрейма (10,2%).  

Пропуски могут быть вызваны:  
1. Отсутствием этих данных для конкретных клиентов; 
2. Программной ошибкой, обнуляющей данные о трудовом стаже при нулевом доходе в настоящее время и наоборот;
3. Ошибкой при заполнении данных.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Общий трудовой стаж в днях и доход в рублях — количественные переменные.  
Пропуски в таких переменных заполняют характерными значениями. Это значения, характеризующие состояние выборки, — набора данных, выбранных для проведения исследования. Чтобы примерно оценить типичные значения выборки, годится медиана.  
Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства, как в нашем случае. Когда в выборке присутствуют выдающиеся значения (например аномальное значение 366413.652744 дней стажа против 578.751554), лучше использовать медиану.  
На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца income_type. 

### Шаг 4: общий вывод.

**Целью данного пректа было исследование надёжности заёмщиков.**  

**Входные данные от банка** — статистика о платёжеспособности клиентов.  
Данные включают информацию о возрасте, поле, уровне образования клиента, общем трудовом стаже и типе занятости, семейном положении и количестве детей в семье, ежемесячном доходе, цели получения кредита, наличии задолженности по возврату кредитов.

**1.** Изучив общую информацию о данных, были выявлены проблеммы в исходных данных. Была проведена **предобработка данных**.
- Удаление пропусков в 10,2% ячеек двух столбцов: days_employed — общий трудовой стаж в днях и total_income — ежемесячный доход. Заменили их медианным значением по каждому типу из столбца income_type - тип занятости;  
- Обработка аномальных значений: отрицательное количество дней трудового стажа в столбце days_employed заменили положительными с помощью метода abs()  
- Удаление строк, в которых встречаются аномальные значения из столбца children (-1 и 20);  
- Замена вещественного типа данных в столбце total_income на целочисленный;  
- Обработка неявных дубликатов в столбце education. Приведение к нижнему регистру значений, записанных по-разному: с использованием заглавных и строчных букв;  
- Удаление строк-дубликатов;  
- Категоризация данных: создание в датафрейме data столбца total_income_category с категориями по уровню дохода и столбца purpose_category, в который вошли  четыре основные категории целей кредита.  

**2. Исследование данных и ответ на вопросы**:  
- Есть ли зависимость между количеством детей и возвратом кредита в срок?  
- Есть ли зависимость между семейным положением и возвратом кредита в срок?  
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?  
- Как разные цели кредита влияют на его возврат в срок?  


**Главные выводы.**  
**1. Бездетные заемщики реже имели задолженность по кредиту, чем заемщики, имеющие 1-2 детей.**  
*Большинство заемщиков бездетные. Учитывая отсутствие дополнительной нагрузки, связанной с иждевенцами, такие заемщики чаще возвращали кредит вовремя. Семьи с одним или двумя детьми возвращали кредит более, чем в 90% случаев. Доля семей с тремя детьми, имевших задолженность по кредиту близка к доле бездетных семей, однако количество таких семей составляет всего 1,5% от всех заемщиков, поэтому эти данные не могут быть достоверными.
Данных по заемщикам с четырьмя и пятью детьми недостаточно для корректных выводов по ним.*
  
**2. Заемщики, которые состоят или когда-либо состояли в зарегистрированном браке надежнее тех, которые не женаты\не замужем.**  
*Самая многочисленная группа заемщиков состоит в браке и имеет процент задолженности по кредиту 7,56%, незначительно превышающий процент задолженности заемщиков в разводе или вдовцов, т.е. тех, кто в прошлом состоял в браке - 7,06%. Наиболее надежными можно считать категорию заемщиков, являющихся вдовцами\вдовами. Процент задолженности по кредиту минимальный и составляет 6,62%. Вероятно людей, вступающих в зарегистрированный брак (в том числе в прошлом), можно считать наиболее ответственными, что положительно сказывается на их способности своевременно исполнять в том числе и финансовые обязательства.
Самой ненадежной группой клиентов являются не женатые\не замужние.*  
  
**3. Кредит в срок чаще возвращают заемщики с доходом до 70000 рублей и доходом от 200001 до 1000000 рублей.**  
*Заемщики с доходом 70001 - 200000 рублей допускают задолженность по кредиту в среднем на 1,5% чаще предыдущей группы.
Этот факт опровергает гипотезу, что клиент с большим доходом более надежный. Вероятно на возврат кредита в значительной степени влияют прочие факторы, и нельзя делать вывод о надежности клиента только по уровню дохода.
Количество заемщиков с доходом более 1000000 настолько незначительно, что невозможно сделать корректные выводы относительно их надежности.*  
  
**4. Возврат кредита в срок зависит от целей кредита.**  
*Наибольшую надежность имеют заемщики, оформляющие кредит на недвижимость.  
Самыми проблемными для целей возврата кредита являются заемщики, оформляющие кредит на операции с автомобилем. Процент таких заемщиков, имевших задолженность по кредиту, максимальный и составляет 9.35%. Вероятно в этом случае оправданным будет оформление страхования займов.  
Займы на получение образования имеют процент задолженности 9.25%, близкий к данным по категории 'операции с автомобилем'. Однако именно в этой категории присутствует наиболее проблемная группа заемщиков, оформивших кредит на получение дополнительного образования.*

**Рекомендации.**  

✔️Портрет идеального заёмщика: вдовец\вдова с доходом от 200 тыс.руб. до 1 млн.руб., не имеющий детей, оформляющий кредит на операции с недвижимостью.  

✔️Хороший заёмщик: человек с доходом от 200 тыс.руб. до 1 млн.руб., состоящий в зарегистрированном браке, не имеющий детей, оформляющий кредит на операции с недвижимостью или человек с доходом от 30 до 50 тыс.руб., состоящий в зарегистрированном браке или разведенный, не имеющий детей, оформляющий кредит на проведение свадьбы.  

❌Плохой заемщик не женат\не замужем, имеющий 2 и более детей с доходом от 50 до 200 тыс.руб., цель кредита - операции с автомобилем. 

Для более точной оценки надёжности заёмщиков можно было бы оценить взаимосвязь количества дней просрочки по кредиту и уровня дохода клиента, а также зависимость типа занятости клиента и возврата кредита в срок.   

